File: python/coin_change.py
class Solution(object):
    # O(n * m) time | O(n) space
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [float("infinity")] * (amount + 1)
        dp[0] = 0

        for c in coins:
            for i in range(1, len(dp)):
                if i >= c and dp[i - c] != float("infinity"):
                    dp[i] = min(dp[i], 1 + dp[i - c])
        return dp[-1] if dp[-1] != float("infinity") else -1


solution = Solution()
# print(solution.coinChange([1, 2, 5], 11))
print(solution.coinChange([2], 3))

File: src/main/java/june_2025/CoinChange.java
package june_2025;

import java.util.Arrays;

public class CoinChange {

    public static void main(String[] args) {
//        int[] coins = {1, 2, 5};
//        int amount = 11;

        int[] coins = {2};
        int amount = 4;

        CoinChange coinChange = new CoinChange();
        int result = coinChange.coinChange(coins, amount);
        System.out.println(result);
    }

    // O(c * a) time | O(a) space
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, Integer.MAX_VALUE);

        dp[0] = 0;
        for (int c : coins) {
            for (int i = 1; i <= amount; i++) {
                if (i >= c) {
                    if (dp[i - c] == Integer.MAX_VALUE) {
                        continue;
                    }
                    dp[i] = Math.min(dp[i], 1 + dp[i - c]);
                }
            }
        }
        return dp[dp.length - 1] == Integer.MAX_VALUE ? -1 : dp[dp.length - 1];
    }

}
